### PR TITLE
LLM: enable chatglm target_model ipex

### PR DESCRIPTION
**Pull request by hzjane**
## Description
enable chatglm target_model ipex

### 1. Why the change?
enable chatglm target_model ipex

### 2. User API changes
No

### 3. Summary of the change 
1. support chatglm ipex n tokens inputs
2. expand ipex kv_cache to bigdl kv_cache
3. ipex optimize before change to bigdl bf16 linear

### 4. How to test?
- [ ] N/A
- [ ] Unit test
- [x] Application test
- [ ] Document test
- [ ] ...

### 5. New dependencies

- [ ] New Python dependencies
       - Dependency1 
       - Dependency2
       - ...
- [ ] New Java/Scala dependencies and their license
       - Dependency1 and license1
       - Dependency2 and license2
       - ...

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 1]
```diff
@@ -518,6 +518,15 @@ def ggml_convert_low_bit(model, qtype, optimize_model=True,
                 f"format......")
     modules_to_not_convert = [] if modules_to_not_convert is None else modules_to_not_convert
 
+    # using ipex optimizer before changing to bigdl linear
+    _enable_ipex = os.getenv("BIGDL_OPT_IPEX")
+    _enable_ipex = (_enable_ipex is not None) and (_enable_ipex.lower() == "true")
+    _enable_ipex = _enable_ipex and (qtype == ggml_tensor_qtype["bf16"])
+    logger.info(f"BIGDL_OPT_IPEX: {_enable_ipex}")
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/transformers/convert.py:525]: We don't need to log on GPU and other confusing situations?

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 2]
```diff
@@ -131,62 +131,100 @@ def clear_benchmarks(self):
...
+                key = key.expand(-1, -1, query_group_size, -1, -1)
+                key = key.contiguous().view(len1, len2 * query_group_size,
+                                            len0,  len3).permute(2, 0, 1, 3)
+                value = value.permute(1, 2, 0, 3).unsqueeze(-3)
+                value = value.expand(-1, -1, query_group_size, -1, -1)
+                value = value.contiguous().view(len1, len2 * query_group_size,
+                                                len0, len3).permute(2, 0, 1, 3)
+                list = []
+                list.append(key[:cur_len, :, :, :])
+                list.append(value[:cur_len, :, :, :])
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/transformers/speculative.py:156]: list = [key[:cur_len, :, :, :], value[:cur_len, :, :, :]]

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 3]
```diff
@@ -131,62 +131,100 @@ def clear_benchmarks(self):
...
-        else:
-            k0 = torch.ones(len0, len1, len2 + max_new_tokens, len3,
-                            dtype=torch.float32)
-            v0 = torch.ones(len0, len1, len2 + max_new_tokens, len3,
-                            dtype=torch.float32)
-            past_key_values_storage.append((k0, v0))
-            if not _enable_ipex:
+            ipex_past_key_values = [
+                [pkv[1].permute(1, 2, 0, 3)[:, :, :cur_len, :],
+                    pkv[2].permute(1, 2, 0, 3)[:, :, :cur_len, :]]
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/transformers/speculative.py:161]: format

**Review by glorysdj — APPROVED** [review]
LGTM

**Pull request merged**